Make package-compatibility and package-fetcher stateless

voxsim opened the pull request:
**Summary**
Make package-compatibility and package-fetcher stateless and indipendent from package-resolver.
I made some ugly choices, but imho the road towards https://github.com/yarnpkg/yarn/issues/2918 is open :D (I hope this merge it will not close that issue).
I think it is important to be incremental as possible because I am touching the core O.o
Big Pro: now we can time every part and maybe introduce some unit tests :)

**Test plan**
Tests are green, before made every change I carefully check if every line were covered indirectly from tests.

arcanis commented on src/package-resolver.js:
```diff
@@ -300,25 +291,25 @@ export default class PackageResolver {
+}
 
-    return collapseToPattern;
+return collapseToPattern;
```
Indentation's broken 😃

arcanis commented:
That's really nice! What would you think about moving `Config` inside the function parameters as well? So, instead of:

```js
export class {

    constructor(config) {
        this.config = config;
        this.fetcher = new PackageFetcher(this.config)
    }

    doSomething() {
        return this.fetcher.doSomething();
    }

}
```

We could have:

```js
export class {

    doSomething({ config }) {
        return PackageManager.doSomething(..., { config });
    }

}
```

voxsim commented on src/package-fetcher.js:
```diff
@@ -75,10 +72,13 @@ export default class PackageFetcher {
 
-  async init(): Promise<void> {
-    let pkgs = this.resolver.getPackageReferences();
+  async init(pkgs: Array<Manifest>): Promise<Array<Manifest>> {
```
I think the right name is `fetch` :) what do you think?

arcanis replied:
👍

voxsim commented:
mmm @arcanis please see my changes: I prefer to export multiple functions instead of one class.
For now we need *always* the config as parameter because we use the report, etc.. it is too complicated to make it as an option imho.

arcanis commented:
As you wish. I think a class would be better because it would be easier to type it, and it would make it possible to easily pass it around (for example, at some point we can imagine the resolver would need to have access to the fetcher in order to extract the correct information from a git repository), but that's something we can change later if we need to.

This PR looks fine by me, the only thing that worries me a bit is that it will introduce a delta between parts of the architecture (most notably the resolver, which will still use the `new PackageResolver()` architecture while the fetchers will be quite different). Do you plan to also work on the other tools as followups?

voxsim commented:
@arcanis yep i am trying to working on other tools incrementally, but they are very difficult to refactor..
One other option is that I can continue to work on this pr, but it will be difficult to merge later maybe O.o
Let me know what you suggest to do :)

voxsim commented:
And if you prefer I can reintroduce classes, I am open to revert my last experiment in case :P

bestander commented:
Ping @arcanis, good to merge?

voxsim commented:
I added some commits to reduce some confusions: i removed seedPatterns and newPatterns. Now in the manifest there is a flag called `fresh` that tell you if the pattern is new. I am starting to understand the internal flow of package-resolver :)

arcanis commented on src/types.js:
```diff
@@ -125,6 +125,8 @@ export type Manifest = {
 
   workspaces?: Array<string>,
+
+  fresh?: boolean
```
Could you add a small comment here to explain what means `fresh` in this context, and a short summary detailing how it is used inside the Yarn internals?

voxsim commented:
Mmh when I program I usually type `yarn build && yarn test` to check if everything is ok after rebasing. There was an issue in the package.json that every time we run `yarn lint` we run `yarn run prettier` instead of `yarn run lint-prettier`, the main issue is that I prettify some random files from last commits O.o and maybe we don't notice because we don't lint correctly and stop the pipeline in case.

voxsim commented:
Btw I committed all the prettified files in one commit (the last one). If there are any problems I can revert it without problems.

arcanis commented:
I'm not familiar with prettier - what's the different? If I understand correctly, `yarn prettier` was updating the file, whereas `yarn lint-prettier` will only be reporting warnings, letting us manually run prettier, is that it?

voxsim commented:
Yep the issue is that if we run `yarn run prettier` in the lint phase we updated files in place and we never commit :(

arcanis commented:
Ok 👍 I'll wait for CI checks to end, and then I'll merge everything

voxsim commented:
Mhh all failed tests seem flaky O.o @arcanis if you catch something please let me know

bestander commented:
Needs a rebase anyway :)

voxsim commented:
Done, I hope everything will be alright :P
